scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,531 @@
+    pub fn new(session: &Arc<Session>) -> GenerationFetcher {
+        GenerationFetcher {
+            generations_table_name: "system_distributed.cdc_generation_timestamps".to_string(),
+            streams_table_name: "system_distributed.cdc_streams_descriptions_v2".to_string(),
+            session: Arc::clone(session),
+        }
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_all_stream_generations_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps';
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    pub async fn fetch_all_generations(&self) -> Result<Vec<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_all_stream_generations_query());
+        self.select_consistency(&mut query).await?;
+        query.set_page_size(DEFAULT_PAGE_SIZE);
+
+        let mut rows = self
+            .session
+            .query_iter(query, &[])
+            .await?
+            .into_typed::<(chrono::Duration,)>();
+
+        let mut generations = vec![];
+
+        while let Some(next_row) = rows.next().await {
+            let (generation,) = next_row?;
+            generations.push(Generation(generation));
+        }
+
+        Ok(generations)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_generation_by_timestamp_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time <= ?
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Given a timestamp of an operation, fetch generation that contains this operation.
+    pub async fn fetch_generation_by_timestamp(
+        &self,
+        time: &chrono::Duration,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_generation_by_timestamp_query());
+        self.select_consistency(&mut query).await?;
+
+        let result = self.session.query(query, (Timestamp(*time),)).await?.rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_generation_by_timeuuid_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time <= dateOf( ? )
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Given a timeuuid of an operation, fetch generation that contains this operation.
+    pub async fn fetch_generation_by_timeuuid(
+        &self,
+        time: &Uuid,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_generation_by_timeuuid_query());
+        self.select_consistency(&mut query).await?;
+
+        let result = self.session.query(query, (time,)).await?.rows;
+
+        GenerationFetcher::return_single_row(result)
```

Comment:
Let's not duplicate the logic of `fetch_generation_by_timestamp`. You can extract a timestamp from the timeuuid and pass it to the `fetch_generation_by_timestamp` function.

I think the extraction can be done in the following way:

```rust
let (secs, nanos) = time
    .to_timestamp()
    .ok_or("Uuid is not a timeuuid")?
    .to_unix();
let timestamp =
    chrono::Duration::seconds(secs as i64) + chrono::Duration::nanoseconds(nanos as i64);
```